### PR TITLE
tests: benchmarks: multicore: Check PWM signal in S2RAM with PWM

### PR DESCRIPTION
**Pull request by nordic-segl**
1. tests: benchmarks: multicore: idle_with_pwm: Fix device runtime PM

API for PWM doesn't provide function for enabling/disabling the device.
As a result, device runtime PM doesn't know when it shall wake up the device (get) / put device to low power operation (put).
Therefore, user must manually call:
- pm_device_runtime_get(&pwm_device);
- pm_device_runtime_put(&pwm_device);

Fix PWM device runtime Power Management test.

2. tests: benchmarks: multicore: Rename idle_with_pwm to idle_pwm_led

Rename test idle_with_pwm to idle_pwm_led to match naming convention.

3. tests: benchmarks: multicore: Check PWM signal in S2RAM with PWM

Add idle_pwm_loopback that checks if PWM signal is present - count rising/falling edges using input GPIO.

### CODE REVIEW

**Review by nordic-piks — APPROVED** [review]

**Comment by peknis** [comment]
If you intend to remove the document, you might need to have a changelog entry. Either in Other samples or Documentation section. And to fix the doc build failure, remove the link to the test doc from the release-notes-2.8.0.rst

**Comment by nordic-segl** [comment]
> If you intend to remove the "PWM in low power states test" document, you might need to have a changelog entry. Either in Other samples or Documentation section. And to fix the doc build failure, remove the link to the test doc from the release-notes-2.8.0.rst

I added GPIOTE to count rising/falling edges (in order to confirm that PWM signal is generated).
However, that blocks entering low power states.
First, I need to "hack" this issue somehow. Then, I will decide what to do with the documentation.

**Comment by nordic-segl** [comment]
I got following results (all values in mA):
**idle_pwm_led**
idle_pwm_led.nrf54h20dk_cpuapp_cpurad.idle
0.454232 / 0.021468

idle_pwm_led.nrf54h20dk_cpuapp_cpurad.idle_fast
0.749777 / 0.021325

idle_pwm_led.nrf54h20dk_cpuapp_cpurad.s2ram
0.45446 / 0.003085

idle_pwm_led.nrf54h20dk_cpuapp_cpurad.s2ram_fast
0.753721 / 0.003206

**idle_pwm_loopback**
idle_pwm_loopback.nrf54h20dk_cpuapp_cpurad.idle
1.046599 / 0.021773

idle_pwm_loopback.nrf54h20dk_cpuapp_cpurad.idle_fast
1.197287 / 0.021791

idle_pwm_loopback.nrf54h20dk_cpuapp_cpurad.s2ram
1.037218 / 0.003108

idle_pwm_loopback.nrf54h20dk_cpuapp_cpurad.s2ram_fast
1.195696 / 0.003104

In IDLE, current is ~0.021791 no matter what test (led/loopback) nor PWM instance (normal/fast).
In S2RAM, current is ~0.003108 no matter what test (led/loopback) nor PWM instance (normal/fast).
Thus, entering low power state seems to work correctly.

In pwm_led test, when cores are active, current consumption is ~0.454232 on normal pwm instance.
And, ~0.749777 on fast pwm instance.
Value for fast instance is higher by ~0.3 mA than value for slow instance. (This is a bit suspicious because in pwm_loopback test current difference is only 0.16 mA.)
Current consumption in active state is similar no matter if core woke up from IDLE or S2RAM.

In pwm_loopback test, when cores are active, current consumption is ~1.040000 on normal pwm instance.
And, ~1.195696 on fast pwm instance.

**Comment by nordic-piks** [comment]
> @nordic-piks : How to handle redundancy in CI? I think, in power consumption Jenkins job, it's enough to run idle_pwm_loopback. This test will confirm power consumption in low power states. Also, it will fail when there is no PWM signal on output.

Redundancy will occur only at nightly build (full scope), not at integration. I think there is nothing to worry about, just leave as it is.

**Comment by nordic-segl** [comment]
Please review.

**Review by nordic-pikr — APPROVED** [review]

File: tests/benchmarks/multicore/idle_pwm_loopback/Kconfig [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+config TEST_SLEEP_DURATION_MS
+	int "Amount of time (in miliseconds) the core is sleeping"
```
Comment (tejlmand) [inline tests/benchmarks/multicore/idle_pwm_loopback/Kconfig:8]: Minor nit, not blocking.

Perhaps shorten this to:
```suggestion
	int "Core sleep duration (miliseconds)"
```
after all this is a title and details are in the help text.

File: tests/benchmarks/multicore/idle_pwm_loopback/Kconfig.sysbuild [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+source "${ZEPHYR_BASE}/share/sysbuild/Kconfig"
+
+config REMOTE_BOARD
+	string "The board used for remote target"
```
Comment (tejlmand) [inline tests/benchmarks/multicore/idle_pwm_loopback/Kconfig.sysbuild:10]: from where does this pattern of having a prompt here originates ?

The exact same construct was commented here: https://github.com/nrfconnect/sdk-nrf/pull/19018#discussion_r1853704191
and I would like to get this construct removed from the root source where people copy things.

Please adjust in similar fashion as describe in this comment: https://github.com/nrfconnect/sdk-nrf/pull/19018#discussion_r1853704191

File: tests/benchmarks/multicore/idle_pwm_loopback/remote/Kconfig [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+config TEST_SLEEP_DURATION_MS
+	int "Amount of time (in miliseconds) the core is sleeping"
```
Comment (tejlmand) [inline tests/benchmarks/multicore/idle_pwm_loopback/remote/Kconfig:8]: Minor nit, not blocking.

Perhaps shorten this to:
```suggestion
	int "Core sleep duration (miliseconds)"
```
after all this is a title and details are in the help text.

File: tests/benchmarks/multicore/idle_pwm_loopback/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    harness: pytest
+    harness_config:
+      fixture: spi_loopback
+      pytest_root:
+        - "${CUSTOM_ROOT_TEST_DIR}/test_measure_power_consumption.py::test_measure_and_data_dump_pwm_and_idle"
+
+  benchmarks.multicore.idle_pwm_loopback.nrf54h20dk_cpuapp_cpurad.s2ram:
+    tags: ppk_power_measure
+    extra_args:
+      SB_CONF_FILE=sysbuild/nrf54h20dk_nrf54h20_cpurad.conf
```
Comment (tejlmand) [inline tests/benchmarks/multicore/idle_pwm_loopback/testcase.yaml:40]: can be removed if following the remote board suggestion.
https://github.com/nrfconnect/sdk-nrf/pull/19018#discussion_r1853707524

File: tests/benchmarks/multicore/idle_pwm_loopback/sysbuild/nrf54h20dk_nrf54h20_cpurad.conf [hunk 5]
```diff
@@ -0,0 +1 @@
+SB_CONFIG_REMOTE_BOARD="nrf54h20dk/nrf54h20/cpurad"
```
Comment (tejlmand) [inline tests/benchmarks/multicore/idle_pwm_loopback/sysbuild/nrf54h20dk_nrf54h20_cpurad.conf:1]: Please remove and follow the remote board suggestion.
See also: https://github.com/nrfconnect/sdk-nrf/pull/19018#discussion_r1853707129

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please cleanup the remote board handling.

You may take a look at how this was done in https://github.com/nrfconnect/sdk-nrf/pull/19018

File: tests/benchmarks/multicore/idle_pwm_loopback/Kconfig (hunk 1 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_loopback/Kconfig:8]: I wish to keep this as it is. It's already in several places:
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_ppr/Kconfig#L8
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_spim_loopback/Kconfig#L8
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_stm/Kconfig#L8
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_wdt/Kconfig#L8

File: tests/benchmarks/multicore/idle_pwm_loopback/Kconfig.sysbuild (hunk 2 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_loopback/Kconfig.sysbuild:10]: It's all over the place :smile:
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_adc/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_clock_control/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_comp/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_gpio/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_hpu_temp_meas/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_outside_of_main/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_spim/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_twim/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_uarte/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_usb/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_wdt/Kconfig.sysbuild#L10
https://github.com/nrfconnect/sdk-nrf/blob/main/tests/benchmarks/multicore/idle_with_pwm/Kconfig.sysbuild#L10

I think, this solution was proposed in the first idle_xxx test and then copied to many others.

File: tests/benchmarks/multicore/idle_pwm_loopback/Kconfig (hunk 1 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_loopback/Kconfig:8]: OK, I will add commit that changes this KConfig title in all files.

File: tests/benchmarks/multicore/idle_pwm_loopback/remote/Kconfig (hunk 3 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_loopback/remote/Kconfig]: Fixed

**Comment by nordic-segl** [comment]
@tejlmand : See last commit titled "tests: benchmarks: multicore: idle_pwm_*: Rework SB_CONFIG_REMOTE_BOARD"
Is this what You meant? If yes, I can apply this to other tests I have authored.

File: tests/benchmarks/multicore/idle_pwm_loopback/sysbuild/nrf54h20dk_nrf54h20_cpurad.conf (hunk 5 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_loopback/sysbuild/nrf54h20dk_nrf54h20_cpurad.conf]: removed

File: tests/benchmarks/multicore/idle_pwm_loopback/testcase.yaml (hunk 4 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_loopback/testcase.yaml]: All occurrences of `SB_CONF_FILE=sysbuild/nrf54h20dk_nrf54h20_cpurad.conf` were removed from `extra_args`.

File: tests/benchmarks/multicore/idle_pwm_led/Kconfig.sysbuild [hunk 6]
```diff
@@ -7,4 +7,5 @@
 source "${ZEPHYR_BASE}/share/sysbuild/Kconfig"
 
 config REMOTE_BOARD
-	string "The board used for remote target"
+	string
+	default "$(BOARD)/$(SOC)/cpurad" if SOC_NRF54H20_CPUAPP
```
Comment (nordic-segl) [inline tests/benchmarks/multicore/idle_pwm_led/Kconfig.sysbuild:11]: @tejlmand : Or shall I use
default "$(BOARD)/**nrf54h20**/cpurad" if SOC_NRF54H20_CPUAPP
as in https://github.com/nrfconnect/sdk-nrf/pull/19018#discussion_r1853704191?

File: tests/benchmarks/multicore/idle_pwm_loopback/Kconfig.sysbuild (hunk 2 above)
Comment (tejlmand) [reply tests/benchmarks/multicore/idle_pwm_loopback/Kconfig.sysbuild]: > This solution seems to have one advantage - it allows to select different remote core for the same 'build" core. For example:

and if that is a need, then I have a solution for that as well :wink:

File: tests/benchmarks/multicore/idle_pwm_led/Kconfig.sysbuild (hunk 6 above)
Comment (tejlmand) [reply tests/benchmarks/multicore/idle_pwm_led/Kconfig.sysbuild:11]: the `$(SOC)` is not defined as a variable you can use in Kconfig, so the actual expansion of this line becomes:
`nrf54h20dk//cpurad`.

So at least you should remove `$(SOC)` as it gives the wrong impression that this is a valid var.

The reason that `nrf54h20dk//cpurad` is also a valid board target when the board is single SoC.

**Comment by tejlmand** [comment]
> Is this what You meant? 

Yes it was.

> If yes, I can apply this to other tests I have authored.

:+1:

File: tests/benchmarks/multicore/idle_pwm_led/Kconfig.sysbuild (hunk 6 above)
Comment (nordic-segl) [reply tests/benchmarks/multicore/idle_pwm_led/Kconfig.sysbuild:11]: Changed to `default "$(BOARD)/nrf54h20/cpurad" if SOC_NRF54H20_CPUAPP`

**Review by tejlmand — APPROVED** [review]

**Comment by nordic-segl** [comment]
DNM until changes to `SB_CONFIG_REMOTE_BOARD` are applied in other tests too.
